### ISSUE TITLE
PSA: removing role levels

### ISSUE THREAD

**Issue by sausin** [JosephSilber/bouncer#223]
Hi,

I noticed the table for `roles` had column entry for `level`. I had asked @JosephSilber about this and he mentioned that this was planned to be removed, and I agree that it probably has no use case.

The intent of opening issue is to document the removal of this feature and any reasons why people might think that they need `level` (but they really don't :wink:). I fully expect that this `issue` will be promptly closed once feedback is collected :smile: 

In my simple implementation in a project I had something similar to enable access to the highest level a user had access to. Some other reasons may exist and I will try to list cases that I can image. If others have any other cases, it would be great to have them documented before the feature is up for removal.

### Scenario 1 - Action Control

A user has permissions to perform certain actions based on his/her access level. If a check is made for the higher level the user has and if it's found to be lower than the required access level, access is denied. If the user's highest role is greater than or equal to required, access is granted.

*Why this is not required with bouncer?*

Every action can be defined as an ability. If the user has access to an ability, then he / she is allowed. This is much simpler!

### Scenario 2 - Route Control

Similar to the previous one but this time if access control to certain routes is required and anyone with a higher than certain access level should be allowed. If the test is truthy, access is granted.

*Why this is not required with bouncer?*

This can be checked and defined as an ability or as a role. It is also possible to assign multiple roles so there is no restriction. In fact I think having multiple roles is better as certain roles can be removed while leaving other access control functionality intact. Again, there is no need to have 'level' as a direct check on role can be made.

### Any other reasons?

Looking forward to getting reasons from other people. It would be even more awesome if someone can explain why this might really be _required_.

**Comment by sausin** [comment]
Okay maybe one reason in favor of having levels. If a project needs everyone with an access level to be able to provider "_lower_" access level access to others then 'level' setting could be handy.

I can see the implementation with a 'level' as not being so complicated right now. It is still possible to do this without levels, but then a lot of abilities would need to be defined and checked

**Comment by JosephSilber** [comment]
We're not looking to add any _new_ functionality to levels. We're only thinking of maybe removing the current functionality, since it's no longer that useful, and adds a lot of complexity to the query for getting abilities.

The way levels work currently is that if a user has a role with a level assigned to them, they're also allowed all abilities with a lower level than the one they have. For example:

```php
$user = User::first();

$subscriber = Bouncer::role()->create([
    'name' => 'subscriber',
    'level' => 1,
]);

$admin = Bouncer::role()->create([
    'name' => 'admin',
    'level' => 2,
]);

Bouncer::allow($subscriber)-to('view', Post::class);
Bouncer::assign($admin)->to($user);

$user->can('view', Post::class); // true
```

In the example above, the user may view posts, even though admins have never explicitly been allowed to view posts. But since subscribers can view posts, and subscribers are at a lower level than admins, we allow the user with an admin role to view posts.

---

The reason we're thinking about removing them is because:

1. It adds a lot of complexity to the query, and has even produced some bugs when used with different DB drivers.

2. The way levels works is very confusing. Should roles with a _lower_ level be granted abilities allowed for roles with _higher_ levels, or the other way around? Currently a role is allowed all abilities of roles with a _lower_ level, but that is an arbitrary decision, and leads to a lot of uncertainty.

3. If someone really needs hierarchical levels, they probably want to [have a full tree](https://github.com/JosephSilber/bouncer/issues/201#issuecomment-326963986), where there can be multiple top-levels, and each can have their own tree of roles that can each have as many children at the same level as needed. We currently don't have that, and adding that would add even more complexity to all this, for little gain.

4. Now that we have more general abilities, such as:

    ```php
    // Allowing everything
    Bouncer::allow('admin')->everything();

    // Allowing owning everything
    Bouncer::allow('admin')->toOwnEverything();

    // Allowing owning a particular type of model
    Bouncer::allow('admin')->toOwn(Client::class);

   // etc.
   ```

    ...the need for levels has been greatly diminished.


---

In closing, can anyone articulate a non-hypothetical need for levels as they currently stand?

**Comment by sausin** [comment]
@JosephSilber Thanks for noting all the details! 

In the example above, I was thinking of having it a implemented in a bit more simple manner, such that the `level` can just be a property on the model without it having any built-in logic around it.

As an example, if we have the following roles (`level` is mentioned in brackets):

* `role-admin` (10)
* `role-higher-user` (9)
* `role-lower-user` (8)
* `role-low-user` (7)

where role `level 10` is the highest role and role `level 7` is the lowest. Someone with role `level 8` can assign someone else a role `level 7` or `level 8` but not `level 9` and so on. The choice on what is allowed depends on the application logic and is independent of the package as such.

To achieve this functionality, one can manually perform a check in the application logic that the person giving someone else a role needs a higher `level` of role. Having this possibility also allows someone using the package to decide for themselves what counts as a _higher_ role.

In short, just having the `level` property remain would be sufficient to achieve this and the logic that you've explained above could still be removed. There are ways to have something that facilitates this built into the package but it may be far too specific for it to be present.

Hope this helps!

**Comment by devinfd** [comment]
I'm really hoping this is **NOT** removed. Role levels was one of the important features that was required for my application and drove me to use Bouncer. We use it for the reason provided above by @JosephSilber. Where an admin has an authorization for an ability already given to a user. This makes for a simpler UI for my accounts to manage there users abilities. 

ps: This is my gut reaction to this proposal. Perhaps I am wrong and need to reevaluate but unless there is another way to "_mock_" the ability to have levels then I am against this proposal.

**Comment by JosephSilber** [comment]
@devinfd the UI itself does not have to become more complex. Assuming we remove role levels from Bouncer, this can still be easily handled in your code whenever a role is edited.

Here's a short example:

```php
public function update(Role $role)
{
    $existing = $role->abilities()->pluck('id');
    $updated = collect(request('abilities_ids'));

    $adding = $updatedAbilityIds->diff($existingAbilityIds);
    $removing = $existingAbilityIds->diff($updatedAbilityIds);

    // Add the abilities to the role, and all with a higher level
    Bouncer::role()->where('level', '<', $role->level)->push($role)
        ->forEach(function ($role) use ($adding) {
            Bouncer::allow($role)->to($adding);
        });

    // Remove the abilities from the role, and all with a lower level
    Bouncer::role()->where('level', '>', $role->level)->push($role)
        ->forEach(function ($role) use ($removing) {
            Bouncer::allow($role)->to($removing);
        });
}
```

I understand that this is more code than was needed before. This however lets us simplify Bouncer quite a bit while still allowing you to keep your elegant UI .

**Comment by JosephSilber** [comment]
I've decided to keep them for now.

**Comment by maartenSnow** [comment]
@sausin  I've been thinking of making something similair. Makes it a lot easier to let different types of users enter the userManagement and asign roles.. 
Although If it's only for application logic, I don't see why the package should support this. I think we are better off making a second migration to add a column? To eep things separated with the package logic.
